Gitlab and e2e tests fix

chmouel opened the pull request:
# Changes

- Gitlab: Only create prun when a MR update with push happen
  We were creating a pipelinerun when an MR was updated for every event,
  which mean every actions like adding labels or setting a status will
  create a new PipelineRun.

  We only want it to get it kicked when it is created or reopen or when a
  there is a commit push attached to that update.

- Move scm.go out of gitea to its own package

  We use those functions to push to Git forge via git directly instead of API which sometime behave weirdly (ie: for gitlab).
  It makes it easier in general for common code and debugging too.

- Tests: Don't cleanup tests when TEST_NOCLEANUP is set

  This let us debug the state of the test after it fails easily.

- Gitea: Make sure we check for the exact number of status

  And make the tests a bit more robust in general.
  we add some timeout to avoid some of the annoying double events

- [E2E Test] Fix wait.Success
  
  We were not checking the last status of the repository and we were never
  checking the number of status we were expecting.

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

# Submitter Checklist

- [ ] 📝 A good commit message is important for other reviewers to understand the context of your change. Please refer to [How to Write a Git Commit Message](https://cbea.ms/git-commit/) for more details how to write beautiful commit messages. We rather have the commit message in the PR body and the commit message instead of an external website.
- [ ] ♽ Run `make test` before submitting a PR (ie: with [pre-commit](https://pipelinesascode.com/dev/tools), no need to waste CPU cycle on CI. (or even better install [pre-commit](https://pre-commit.com/) and do `pre-commit install` in the root of this repo).
- [ ] ✨ We heavily rely on linters to get our code clean and consistent, please ensure that you have run `make lint` before submitting a PR. The [markdownlint](https://github.com/DavidAnson/markdownlint) error can get usually fixed by running `make fix-markdownlint` (make sure it's installed first)
- [ ] 📖 If you are adding a user facing feature or make a change of the behavior, please verify that you have documented it
- [ ] 🧪 100% coverage is not a target but most of the time we would rather have a unit test if you make a code change.
- [ ] 🎁 If that's something that is possible to do please ensure to check if we can add a e2e test.
- [ ] 🔎 If there is a flakiness in the CI tests then don't *necessary* ignore it, better get the flakyness fixed before merging or if that's not possible there is a good reason to bypass it. (token rate limitation may be a good reason to skip).
and make the tests a bit more robust in general.
we add some timeout to avoid some of the annoying double events.

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel committed:
Gitea: Make sure we check for the exact number of status

and make the tests a bit more robust in general.
we add some timeout to avoid some of the annoying double events.

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel committed:
Tests: Don't cleanup tests when TEST_NOCLEANUP is set

This let us debug the state of the test after it fails easily.

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel committed:
Only create prun when a MR update with push happen

We were creating a pipelinerun when an MR was updated for every event,
which mean every actions like adding labels or setting a status will
create a new PipelineRun.

We only want it to get it kicked when it is created or reopen or when a
there is a commit push attached to that update.

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel committed:
Move scm.go out of gitea to its own package

We use those functions to push to Git forge via git directly instead of API which sometime behave weirdly (ie: for gitlab).
It makes it easier in general for common code and debugging too.

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel commented on pkg/provider/gitlab/detect.go:
```diff
@@ -37,7 +37,12 @@ func (v *Provider) Detect(req *http.Request, payload string, logger *zap.Sugared
-		if provider.Valid(gitEvent.ObjectAttributes.Action, []string{"open", "update", "reopen"}) {
+		// on a MR Update only react when there is Oldrev set, since this means
+		// there is a Push of commit in there
+		if gitEvent.ObjectAttributes.Action == "update" && gitEvent.ObjectAttributes.OldRev != "" {
```
@sm43 seems like the OldRev do the trick...

chmouel commented on test/pkg/gitea/test.go:
```diff
@@ -235,9 +245,12 @@ func WaitForStatus(t *testing.T, topts *TestOpts, ref, forcontext string, onlyla
 		}
 		topts.ParamsRun.Clients.Log.Infof("Number of gitea status on PR: %d/%d", numstatus, checkNumberOfStatus)
-		if numstatus >= checkNumberOfStatus {
+		if numstatus == checkNumberOfStatus {
```
this make the gitea e2e more stricts!

piyush-garg commented:
lgtm

savitaashture commented on test/gitea_access_control_test.go:
```diff
@@ -167,8 +169,9 @@ func TestGiteaACLOrgAllowed(t *testing.T) {
 		},
-		NoCleanup:    true,
-		ExpectEvents: false,
+		NoCleanup:            true,
```
Do we want to keep this true by default ? 
asking because it will not cleanup if its true

chmouel replied:
it's a old field not used, but that's not the point of this pr, you can send a cleanup pr if you want later on.

chmouel replied:
removed them as part of a commit

savitaashture commented on test/gitlab_merge_request_test.go:
```diff
@@ -42,25 +45,48 @@ func TestGitlabMergeRequest(t *testing.T) {
+	assert.NilError(t, err)
+	scmOpts.BaseRefName = targetRefName
+	scm.PushFilesToRefGit(t, scmOpts, entries)
+	wait.Succeeded(ctx, t, runcnx, opts, "Merge Request", targetNS, 4, "", mrTitle)
```
I think this won't work 

because https://github.com/chmouel/pipelines-as-code/blob/13a970a2078f4ec8fc896b44efb1c3114d649e39/test/pkg/wait/check.go#L35 the for loop executes only when `numberofprmatch` is `1` :grin:

chmouel replied:
what do you mean it won't work?

chmouel replied:
ah yeah that for loop has never worked, you are right!

chmouel replied:
I fixed this in another commit on this PR

chmouel committed:
[E2E Test] Fix wait.Success

We were not checking the last status of the repository and we were never
checking the number of status we were expecting.

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>
